src/main/java/picard/illumina/ExtractIlluminaBarcodes.java

```diff
@@ -798,16 +915,25 @@ public static BarcodeMatch findBestBarcodeAndUpdateMetrics(final byte[][] readSu
          */
         private static int countMismatches(final byte[][] barcodeBytes, final byte[][] readSubsequence, final byte[][] qualities, int minimumBaseQuality) {
             int numMismatches = 0;
-            // Read sequence and barcode length may not be equal, so we just use the shorter of the two
+
             for (int j = 0; j < barcodeBytes.length; j++) {
-                final int basesToCheck = Math.min(barcodeBytes[j].length, readSubsequence[j].length);
-                for (int i = 0; i < basesToCheck; ++i) {
-                    if (!SequenceUtil.isNoCall(readSubsequence[j][i])) {
-                        if (!SequenceUtil.basesEqual(barcodeBytes[j][i], readSubsequence[j][i])) ++numMismatches;
-                        else if (qualities != null && qualities[j][i] < minimumBaseQuality) ++numMismatches;
+                for (int i = 0; i < barcodeBytes[j].length; ++i) {
+                    if (readSubsequence[j].length <= i) {
+                        break;
+                    }
+                    if (SequenceUtil.isNoCall(readSubsequence[j][i])) {
+                        continue;
+                    }
+                    if (!SequenceUtil.basesEqual(barcodeBytes[j][i], readSubsequence[j][i])) {
```

Comment:
if with continue is slightly faster in that if the initial condition is met the second doesn't even need to be evaluated.

I thought this wouldn't make much of a difference (since I figured the compiler would optimize this down to the same bytecode.. However, several rounds of profiling showed a significant difference. I'd be happy to share my timings if you like.